image-prune

```diff
@@ -24,125 +24,88 @@ import argparse
-        if not quiet:
-            sys.stderr.write("Considering images from {0} ({1})\n".format(name, ref))
-        for link in get_image_links(ref, "images"):
-            if ref in considerable:
-                for consider in considerable[ref]:
-                    if link.startswith(consider):
-                        images.add(link)
-            else:
-                images.add(link)
-
-    return images
-
-
-def get_keepers(quiet=False, open_pull_requests=True, offline=False, checkout_only=False):
-    if checkout_only:
-        targets = set()
-    else:
-        targets = get_image_names(quiet, open_pull_requests, offline)
+def git(*args, cwd='.'):
+    return subprocess.check_output(['git', *args], text=True, cwd=cwd)
+
+
+def get_images_in_commit(commit, cwd='.'):
+    """Returns all images pointed to by a symlink in images/ from the given commit"""
+
+    logger.debug('Keeping all images in %s...', commit)
+    for line in git('cat-file', '-p', f'{commit}:images/', cwd=cwd).splitlines():
+        mode, kind, blob, _name = line.split()
+        if mode == '120000' and kind == 'blob':
+            image = git('cat-file', 'blob', blob, cwd=cwd)
+            logger.debug('  - %s', image)
+            yield image
+
+
+def get_images_in_branches(merge_target, *branch_patterns, cwd='.'):
+    """Returns all images introduced by branches, relative to merge_target"""
+
+    # KISS: just get a diff and grab anything that looks like an image update
+    for ref_line in git('for-each-ref', *branch_patterns, cwd=cwd).splitlines():
+        _rev, _kind, ref = ref_line.split()
+        logger.debug('Considering images changed in %s...', ref)
+        for patch_line in git('diff', f'{merge_target}...{ref}', '--', 'images/', cwd=cwd).splitlines():
+            if match := re.search(r'^\+([-0-9a-z]+-[0-9a-f]+.(qcow2|iso))', patch_line):
+                image = match.group(1)
+                logger.debug('  - %s', image)
+                yield image
+
+
+def get_remote_images():
+    """Returns all images currently used by PRs and origin branches on GitHub"""
+
+    api = github.GitHub()
+    logger.debug('Querying open PRs on %s', api.repo)
+    open_prs = [pr['number'] for pr in api.pulls()]
+    logger.debug('Open PRs: %s', open_prs)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        git('init', cwd=tmpdir)
+
+        # We store HEAD as `head`, origin branches as `origin/{name}` and PRs as `pr/{nr}`
+        patterns = {'HEAD:refs/heads/head', 'refs/heads/*:refs/heads/origin/*'}
+        patterns.update(f'refs/pull/{nr}/head:refs/heads/pr/{nr}' for nr in open_prs)
+        quiet = ['--quiet'] if not logger.isEnabledFor(logging.DEBUG) else []
+        git('fetch', *quiet, f'https://github.com/{api.repo}', *patterns, cwd=tmpdir)
+
+        yield from get_images_in_commit('head', cwd=tmpdir)
+        yield from get_images_in_branches('head', 'refs/heads/origin/*', 'refs/heads/pr/*', cwd=tmpdir)
+
+
+def get_local_images(branches=True):
```

Comment:
Not used anywhere; possibly missing fallback for `--offline`? Or just a leftover?

Reply:
Thanks!  Just a leftover.  It proved to be trivial enough to just implement directly in `get_keepers()`.